src/once_cell.rs

```diff
@@ -0,0 +1,800 @@
+///     // a database call.
+///     # Foobar
+/// }
+///
+/// struct LazyFoobar {
+///     inner: OnceCell<Foobar>,
+/// }
+///
+/// impl LazyFoobar {
+///     fn new() -> Self {
+///         Self {
+///             inner: OnceCell::new(),
+///         }
+///     }
+///
+///     async fn load(&self) -> &Foobar {
+///         self.inner.get_or_init(|| async {
+///             very_expensive_initialization().await
+///         }).await
+///     }
+/// }
+/// ```
+pub struct OnceCell<T> {
+    /// Listeners waiting for a chance to initialize the cell.
+    ///
+    /// These are the users of get_or_init() and similar functions.
+    active_initializers: Event,
+    /// Listeners waiting for the cell to be initialized.
+    ///
+    /// These are the users of wait().
+    passive_waiters: Event,
+    /// State associated with the cell.
+    state: AtomicUsize,
+    /// The value of the cell.
+    value: UnsafeCell<MaybeUninit<T>>,
+}
+
+unsafe impl<T: Send> Send for OnceCell<T> {}
+unsafe impl<T: Send + Sync> Sync for OnceCell<T> {}
+
+impl<T> OnceCell<T> {
+    /// Create a new, uninitialized `OnceCell`.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use async_lock::OnceCell;
+    ///
+    /// let cell = OnceCell::new();
+    /// # cell.set_blocking(1);
+    /// ```
+    pub const fn new() -> Self {
+        Self {
+            active_initializers: Event::new(),
+            passive_waiters: Event::new(),
+            state: AtomicUsize::new(UNINITIALIZED),
+            value: UnsafeCell::new(MaybeUninit::uninit()),
+        }
+    }
+
+    /// Tell whether or not the cell is initialized.
+    ///
+    /// This may not always be accurate. For instance, it is possible for
+    /// another thread to initialize the cell between the time when this
+    /// function is called and the time when the result is actually used.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use async_lock::OnceCell;
+    ///
+    /// # futures_lite::future::block_on(async {
+    /// let cell = OnceCell::new();
+    /// assert!(!cell.is_initialized());
+    /// cell.set(1).await;
+    /// assert!(cell.is_initialized());
+    /// # });
+    /// ```
+    pub fn is_initialized(&self) -> bool {
+        self.state.load(Ordering::Acquire) == INITIALIZED
+    }
+
+    /// Get a reference to the inner value, or `None` if the value
+    /// is not yet initialized.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use async_lock::OnceCell;
+    ///
+    /// # futures_lite::future::block_on(async {
+    /// let cell = OnceCell::new();
+    /// assert!(cell.get().is_none());
+    /// cell.set(1).await;
+    /// assert_eq!(cell.get(), Some(&1));
+    /// # });
+    /// ```
+    pub fn get(&self) -> Option<&T> {
+        if self.is_initialized() {
+            // SAFETY: We know that the value is initialized, so it is safe to
+            // read it.
+            Some(unsafe { self.get_unchecked() })
```

Comment:
Isn't there a possiblity of another thread calling `take` between the check and `get_unchecked` here?

Reply:
ah no, `take` takes a mut ref so this can't happen.

Reply:
The idea is, once `state` is set to `INITIALIZED`, only `&mut` access can be used to directly mutate the value.